SYNTHESIZER: Add dump-loop-contracts into loop-contracts synthesizer

qinheping opened the pull request:
Running the loop-contracts synthesizer with the flag `dump-loop-contracts` will output the synthesized contracts as a JSON file, which can be used to annotate back to the source code with Crangler.


<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

jimgrundy reviewed (changes requested):
I'm ok with the changes to the synthesizer, but the changes to c_wranger and the expression type not so much. They do look like improvements/fixes, but they should be factored into their own PRs so they and their consequences can be evaluated without holding up the synthesizer change.

Do please make those other PRs though.

tautschnig commented on src/goto-synthesizer/goto_synthesizer_parse_options.cpp:
```diff
@@ -47,23 +48,64 @@ int goto_synthesizer_parse_optionst::doit()
-    code_contractst contracts(goto_model, log);
-    contracts.apply_loop_contracts(to_exclude_from_nondet_static);
+    // Default output destination is stdout.
+    std::string json_output_str = "stdout";
```
This looked a bit weird to me on first sight, but am I right that this is understood by Crangler?

qinheping replied:
Yes. https://github.com/diffblue/cbmc/blob/develop/src/crangler/c_wrangler.cpp#L616

tautschnig replied:
Would you mind just adding a comment to make sure others also understand this relationship to Crangler.

tautschnig commented:
With all other PRs now merged: can you use expr2c with a custom configuration that sets up true and false as you initially did in the default configuration? I believe this will (after a rebase) make all tests pass.

qinheping committed:
Add dump-loop-contracts into loop-contracts synthesizer

Running the loop-contracts synthesizer with the flag `dump-loop-contracts` will output the synthesized contracts as a JSON file, which can be used to annotate back to the source code with Crangler.
